src/Shared/UnitTests/FileMatcher_Tests.cs

```diff
@@ -1375,6 +1374,54 @@ public void FileEnumerationCacheTakesExcludesIntoAccount()
             }
         }
 
+        [Fact]
+        public void DriveEnumerationWildcardIsObserved()
+        {
+            using (var env = TestEnvironment.Create())
+            {
+                ChangeWaves.ResetStateForTests();
+                env.SetEnvironmentVariable("MSBUILDFAILONDRIVEENUMERATINGWILDCARD", "1");
+                BuildEnvironmentHelper.ResetInstance_ForUnitTestsOnly();
+
+                (string[] fileList, FileMatcher.SearchAction action, string excludeFileSpec) =  FileMatcher.Default.GetFiles(string.Empty, NativeMethodsShared.IsWindows ? @"c:\**\*.cs" : "/**/*.cs");
+                action.ShouldBe(FileMatcher.SearchAction.FailOnDriveEnumeratingWildcard);
+                fileList.ShouldBeEmpty();
+                excludeFileSpec.ShouldBe(string.Empty);
+
+                (fileList, action, excludeFileSpec) = FileMatcher.Default.GetFiles(string.Empty, NativeMethodsShared.IsWindows ? "c:\\**" : "/**");
+                action.ShouldBe(FileMatcher.SearchAction.FailOnDriveEnumeratingWildcard);
+                fileList.ShouldBeEmpty();
+                excludeFileSpec.ShouldBe(string.Empty);
+
+                (fileList, action, excludeFileSpec) = FileMatcher.Default.GetFiles(string.Empty, NativeMethodsShared.IsWindows ? "c:\\\\**" : "/**");
+                action.ShouldBe(FileMatcher.SearchAction.FailOnDriveEnumeratingWildcard);
+                fileList.ShouldBeEmpty();
+                excludeFileSpec.ShouldBe(string.Empty);
+
+                (fileList, action, excludeFileSpec) = FileMatcher.Default.GetFiles(string.Empty, NativeMethodsShared.IsWindows ? "c:\\\\\\\\**" : "/**");
+                action.ShouldBe(FileMatcher.SearchAction.FailOnDriveEnumeratingWildcard);
+                fileList.ShouldBeEmpty();
+                excludeFileSpec.ShouldBe(string.Empty);
+
+                (fileList, action, excludeFileSpec) = FileMatcher.Default.GetFiles(string.Empty, NativeMethodsShared.IsWindows ? @"c:\**" : "/**");
+                action.ShouldBe(FileMatcher.SearchAction.FailOnDriveEnumeratingWildcard);
+                fileList.ShouldBeEmpty();
+                excludeFileSpec.ShouldBe(string.Empty);
+
+                (fileList, action, excludeFileSpec) = FileMatcher.Default.GetFiles(string.Empty, NativeMethodsShared.IsWindows ? @"\**" : "/**");
+                action.ShouldBe(FileMatcher.SearchAction.FailOnDriveEnumeratingWildcard);
+                fileList.ShouldBeEmpty();
+                excludeFileSpec.ShouldBe(string.Empty);
+
+                (fileList, action, excludeFileSpec) = FileMatcher.Default.GetFiles(string.Empty, NativeMethodsShared.IsWindows ? @"c:////////**" : "/**");
+                action.ShouldBe(FileMatcher.SearchAction.FailOnDriveEnumeratingWildcard);
+                fileList.ShouldBeEmpty();
+                excludeFileSpec.ShouldBe(string.Empty);
+
+                ChangeWaves.ResetStateForTests();
```

Comment:
This line won't be called on an assertion failure above. Is it necessary?

If so, it should probably be in a try-finally, or incorporated as part of the `TestEnvironment` disposing pattern.

Reply:
Used try-finally as resetting state is necessary.